src/controllers/private-rest.ts

```diff
@@ -1,14 +1,128 @@
-import { Path, Post, Route, Tags } from "tsoa";
-import { Badge } from "@domain/readModels/Badge";
+import { Body, Delete, Get, Patch, Path, Post, Route, Tags } from "tsoa";
+import type { BadgeHubDataPort } from "@domain/BadgeHubDataPort";
+import { BadgeHubDataPostgresAdapter } from "@db/BadgeHubDataPostgresAdapter";
+import type { Version } from "@domain/readModels/app/Version";
+import type { ProjectSlug } from "@domain/readModels/app/Project";
+import type { DBInsertUser, DBUser } from "@db/models/app/DBUser";
+import type { DBInsertProject } from "@db/models/app/DBProject";
+import type { DBInsertAppMetadataJSON } from "@db/models/app/DBAppMetadataJSON";
 
+interface UserProps extends Omit<DBInsertUser, "id"> {}
+
+interface ProjectProps extends Omit<DBInsertProject, "slug"> {}
+interface ProjectPropsPartial extends Partial<ProjectProps> {}
+interface DbInsertAppMetadataJSONPartial
+  extends Partial<DBInsertAppMetadataJSON> {}
+
+// TODO verify user_name against logged in user
 @Route("/api/v3")
-@Tags("public")
+@Tags("private")
 export class PrivateRestController {
+  public constructor(
+    private badgeHubData: BadgeHubDataPort = new BadgeHubDataPostgresAdapter()
+  ) {}
+
+  /**
+   * Create a new user
+   */
+  @Post("/users/{userId}")
+  public async insertUser(
```

Comment:
- I agree about the raw DB erros being sent back from the API should change. Maybe we can indeed use some TSOA middleware for that. Could you create a ticket for this?
 
- Extra manual validation is not really that necessary I think because TSOA is already checking the input shape of the requests so that they match the typescript types.

- Now I was also wondering whether I should make an extra layer for the DB, but it felt a bit overengineering for now. Idea currently is that the BadgeHubDataPostgresAdapter translates between Badgehub Domain and DB, the Badgehub domain is used by the API. The abstraction is a bit broken because in some places it made sense to reuse the DB insert typings. But if we want extra validation and shielding, I think we can do that in the BadgeHubDataPostgresAdapter.ts or private-rest.ts.

Reply:
**Edit**: now that I wanted to start working on the file uploads, I see it does make sense to to make an separate abstraction layer for the DB.
So I've created a ticket to follow up on your suggestion (#43), and I'll do that before starting on draft file management (#36)